habitat_baselines/il/data/edfe_data.py

```diff
@@ -0,0 +1,220 @@
+import random
+from typing import List
+
+import cv2
+import lmdb
+import numpy as np
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+import habitat
+from habitat.core.simulator import ShortestPathPoint
+
+
+class EDFEDataset(Dataset):
+    """Pytorch dataset for Embodied Q&A's feature-extractor"""
+
+    def __init__(self, env, config, mode="train"):
+        """
+        Args:
+            env (habitat.Env): Habitat environment
+            config: Config
+            mode: 'train'/'val'
+        """
+        self.env = env
+        self.sim = self.env.sim
+        self.config = config.TASK_CONFIG
+        self.mode = mode
+
+        self.episodes = self.get_all_episodes(config)
+
+        self.dataset_path = config.DATASET_PATH
+
+        self.scene_ids = []
+        self.scene_episode_dict = {}
+
+        # preparing a dict that stores list of episodes for each scene
+        for ep in self.episodes:
+            if ep.scene_id not in self.scene_ids:
+                self.scene_ids.append(ep.scene_id)
+                self.scene_episode_dict[ep.scene_id] = [ep.episode_id]
+            else:
+                self.scene_episode_dict[ep.scene_id].append(ep.episode_id)
+
+        self.disk_cache_exists = self.check_cache_exists()
+
+        if not self.disk_cache_exists:
+            try:
+                self.make_dataset_dirs()
+            except FileExistsError:
+                pass
+            """
+            for each scene > load scene in memory > save frames for each
+            episode corresponding to each scene
+            """
+            print("Saving rgb, seg, depth data to database.")
+
+            train_lmdb = lmdb.open(
+                self.dataset_path.format(split="train"),
+                map_size=int(1e11),
+                writemap=True,
+            )
+
+            val_lmdb = lmdb.open(
+                self.dataset_path.format(split="val"),
+                map_size=int(0.5e11),
+                writemap=True,
+            )
+
+            self.train_count = -1
+            self.val_count = -1
+
+            self.train_txn = train_lmdb.begin(write=True)
+            self.val_txn = val_lmdb.begin(write=True)
+
+            for scene in tqdm(list(self.scene_episode_dict.keys())):
+
+                self.load_scene(scene)
+
+                for ep_id in tqdm(self.scene_episode_dict[scene]):
+
+                    episode = next(
+                        ep for ep in self.episodes if ep.episode_id == ep_id
+                    )
+
+                    pos_queue = episode.shortest_paths[0]
+                    random_pos = random.sample(pos_queue, 9)
+                    self.save_frames(random_pos)
+
+            print("EDFE database ready!")
+
+            if self.mode == "train":
+                self.lmdb_txn = self.train_txn
+            elif self.mode == "val":
+                self.lmdb_txn = self.val_txn
+        else:
+            lmdb_env = lmdb.open(self.dataset_path.format(split=self.mode))
```

Comment:
```suggestion
            lmdb_env = lmdb.open(self.dataset_path.format(split=self.mode), readonly=True, lock=False)
```

I have found that some systems need `lock=False` to support concurrent loading from many workers smoothly.

Reply:
Sorry I missed this suggestion somehow. Have committed it now. Thank you